core/order.go

```diff
@@ -1016,27 +1017,71 @@ func quantityForItem(version uint32, item *pb.Order_Item) uint64 {
 }
 
 func (n *OpenBazaarNode) getPriceInSatoshi(paymentCoin, currencyCode string, amount uint64) (uint64, error) {
+	const reserveCurrency = "BTC"
 	if NormalizeCurrencyCode(currencyCode) == NormalizeCurrencyCode(paymentCoin) || "T"+NormalizeCurrencyCode(currencyCode) == NormalizeCurrencyCode(paymentCoin) {
 		return amount, nil
 	}
 
-	wal, err := n.Multiwallet.WalletForCurrencyCode(paymentCoin)
+	var (
+		currencyDict             = repo.LoadCurrencyDefinitions()
+		originCurrencyDef, oErr  = currencyDict.Lookup(currencyCode)
+		paymentCurrencyDef, pErr = currencyDict.Lookup(paymentCoin)
+		reserveCurrencyDef, rErr = currencyDict.Lookup(reserveCurrency)
+	)
+	if oErr != nil {
+		return 0, fmt.Errorf("invalid listing currency code: %s", oErr.Error())
+	}
+	if pErr != nil {
+		return 0, fmt.Errorf("invalid payment currency code: %s", pErr.Error())
+	}
+	if rErr != nil {
+		return 0, fmt.Errorf("invalid reserve currency code: %s", rErr.Error())
+	}
+
+	originValue, err := repo.NewCurrencyValueFromUint(amount, originCurrencyDef)
 	if err != nil {
-		return 0, fmt.Errorf("currency %s not found in multiwallet", paymentCoin)
+		return 0, fmt.Errorf("parsing amount: %s", err.Error())
+	}
+
+	wal, err := n.Multiwallet.WalletForCurrencyCode(reserveCurrency)
+	if err != nil {
+		return 0, fmt.Errorf("%s wallet not found for exchange rates", reserveCurrency)
 	}
 
 	if wal.ExchangeRates() == nil {
 		return 0, ErrPriceCalculationRequiresExchangeRates
 	}
-	exchangeRate, err := wal.ExchangeRates().GetExchangeRate(currencyCode)
+	reserveIntoOriginRate, err := wal.ExchangeRates().GetExchangeRate(currencyCode)
 	if err != nil {
 		return 0, err
 	}
+	originIntoReserveRate := 1 / reserveIntoOriginRate
+	reserveIntoResultRate, err := wal.ExchangeRates().GetExchangeRate(paymentCoin)
+	if err != nil {
+		// TODO: remove hack once ExchangeRates can be made aware of testnet currencies
+		if strings.HasPrefix(paymentCoin, "T") {
+			reserveIntoResultRate, err = wal.ExchangeRates().GetExchangeRate(strings.TrimPrefix(paymentCoin, "T"))
+			if err != nil {
+				return 0, err
+			}
+		} else {
+			return 0, err
+		}
+	}
 
-	formatedAmount := float64(amount) / 100
-	btc := formatedAmount / exchangeRate
-	satoshis := btc * float64(wal.ExchangeRates().UnitsPerCoin())
-	return uint64(satoshis), nil
+	reserveValue, err := originValue.ConvertTo(reserveCurrencyDef, originIntoReserveRate)
+	if err != nil {
+		return 0, fmt.Errorf("converting to reserve: %s", err.Error())
+	}
+	resultValue, err := reserveValue.ConvertTo(paymentCurrencyDef, reserveIntoResultRate)
+	if err != nil {
+		return 0, fmt.Errorf("converting from reserve: %s", err.Error())
+	}
+	result, err := resultValue.AmountUint64()
+	if err != nil {
+		return 0, fmt.Errorf("unable to represent (%s) as uint64: %s", resultValue.String(), err.Error())
+	}
+	return result, nil
```

Comment:
The new behavior for `getPriceInSatoshi(paymentCurrency, listedCurrency, amount)`:
1. Assume reserve currency is `BTC` (this can be flexible later)
2. Get reserve currency's rates vs. all other currencies.
3. Get rates of payment and listed as it relates to the reserve currency.
4. Take amount in `listingCurrency` and `ConvertTo` reserve currency. Then again to `paymentCurrency`. (this consider divisibility within the CurrencyValue as part of the conversion)
5. Attempt to cast it back into int64 (this would still fail for very large values, but it would be protected)

Reply:
I'd like to see a test for this function as it's pretty critical though it requires a good deal of mocking. Maybe this could be broken out into a separate task.

Reply:
I attempted to protect the conversion of the value as much as possible without requiring a mocked API response or exchange rate provider. I think we should replace the exchange rate interface to provide something like:
```
type ExchangeRater interface{
  // returns the rate of native wallet currency to the reserve currencies
  // indentified by normalized currency code string as the key
  RatesByReserveCurrency() map[string]float64
}
```

When the rates are requested by ob-go, it can get the rates for all CurrencyDefinitions in the mainnet dictionary from the local ExchangeRater (which looks much like the currency ExchangeRate provider in the BTC SPV wallet) merged with the results from any wallets which also implement the ExchangeRater interface. Not sure how native vs wallet values get overwrite priority, but this would give us a local exchange rate impl (which would be tested) and the integration test would be easier. Thoughts?

Reply:
The exchange rater would also need to manage multiple pairs. As different wallets might offer different reserve currencies. I imagine the exchange rater cache would look something like `map[string]map[string]float64` or maybe `map[repo.CurrencyCode]repo.ReserveRatesCache`
where the key is the reserve currency (which would be largely populated with BTC values) and the `repo.ReserveRatesCache` is a model which knows the rates for the `repo.CurrencyCode` key requested. Once we have this mapping, we can ignore the reserve currency completely and could just ask for a ConvertTo from LTC->USD and it would know how to handle the rates and find a matching reserve to convert through. 

Concern: Nothing would stop the wallet from injecting bad rates into the ExchangeRater. Would be nice to find some way to prove the rates are legit. (Granted wallets already require a degree of trust, this design would allow a wallet to inject its rates into any transaction to or from this wallet's native currency(s). Before, the wallet could only impact its own blockchain state on behalf of the user. Here, the wallet might impact vendors which have not accepted the buyer's wallet's rate. Or, worse, could make counterparty verification of the order value inconsistent depending on which rates are consumed on either side of the trade.)